tapeagents/optimize.py

```diff
@@ -0,0 +1,62 @@
+"""Module to optimize agents"""
+
+import logging
+import random
+from typing import Callable
+
+from .agent import Agent
+from .dialog_tape import Tape
+from .io import save_agent
+from .llm_function import LLMFunctionNode
+
+logger = logging.getLogger(__name__)
+
+
+def optimize_demos(
+    agent: Agent,
+    good_tapes: list[Tape],
+    n_demos: int,
+    random_sample: int,
+    seed: int,
+    metric_fn: Callable[[list[Tape]], float],
+    run_agent_fn: Callable[[Agent, list[Tape]], list[Tape]],
+) -> Agent:
+    """
+    Try `random_sample` times to `add_demos()` (see below), measure validation set performance, and keep the best agent.
+    """
+    best_agent = agent
+    best_agent_id = 0
+    best_metric = 0
+
+    for i in range(random_sample):
+        # Add demos to the agent with a different seed for each attempt
+        new_agent = add_demos(best_agent, good_tapes, n_demos, seed=seed + i)
+        save_agent(new_agent, f"agent_{best_agent_id}.yaml")
```

Comment:
we don't want a core library function to safe stuff to files

Reply:
Do we prefer to have a param that would save file in this function OR that this function returns the best_agent as well as all the intermediate agents it tried for observability (the caller would be in charge to save agents)?